File: client/src/main/java/de/qabel/desktop/ui/tray/TrayIconMouseAdapter.java
package de.qabel.desktop.ui.tray;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.MouseMotionAdapter;
import java.util.ArrayList;
import java.util.Timer;
import java.util.TimerTask;

class TrayIconMouseAdapter extends MouseAdapter {
    private final JPopupMenu popup;
    private final Runnable showApp;
    private final Runnable bringAppToFront;
    private boolean inBound;
    private Timer notificationTimer;

    TrayIconMouseAdapter(Runnable showApp, Runnable bringAppToFront) {
        popup = buildSystemTrayJPopupMenu();
        this.showApp = showApp;
        this.bringAppToFront = bringAppToFront;

        popup.addMouseMotionListener(new MouseMotionAdapter() {
            @Override
            public void mouseMoved(MouseEvent e) {
                inBound = true;
            }
        });

        popup.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseExited(MouseEvent e) {
                if (isInBounds(e, popup)) {
                    return;
                }
                inBound = false;
            }

            private boolean isInBounds(MouseEvent e, JPopupMenu popup) {
                return e.getX() < popup.getBounds().getMaxX() &&
                    e.getX() >= popup.getBounds().getMinX() &&
                    e.getY() < popup.getBounds().getMaxY() &&
                    e.getY() >= popup.getBounds().getMinY();
            }
        });
    }

    private void hideIfOutOfBounds() {
        if (popup.isVisible() && !inBound) {
            popup.setVisible(false);
            stopTimer();
        }
        inBound = false;
    }

    private synchronized void startTimer() {
        stopTimer();
        notificationTimer = new Timer();
        notificationTimer.schedule(new TimerTask() {
            @Override
            public void run() {
                hideIfOutOfBounds();
            }
        }, 500, 1500);
    }

    private synchronized void stopTimer() {
        if (notificationTimer != null) {
            notificationTimer.cancel();
            notificationTimer = null;
        }
    }

    public void setInBound(boolean inBound) {
        this.inBound = inBound;
    }

    private JPopupMenu buildSystemTrayJPopupMenu() {
        final JPopupMenu menu = new JPopupMenu();
        final JMenuItem exitMenuItem = new JMenuItem("Exit");

        menu.add(exitMenuItem);
        exitMenuItem.addActionListener(ae -> System.exit(0));
        return menu;
    }

    @Override
    public void mouseReleased(MouseEvent e) {
        if (isSingleLeftClick(e)) {
            bringAppToFront.run();
        } else if (isDoubleLeftClick(e)) {
            showApp.run();
        } else if (isRightClick(e)) {
            Point position = calculatePosition(e);
            popup.setLocation(position.x, position.y);
            popup.setVisible(true);
            startTimer();
        }
    }

    private boolean isRightClick(MouseEvent e) {
        return e.getButton() == MouseEvent.BUTTON3;
    }

    private boolean isSingleLeftClick(MouseEvent e) {
        return e.getButton() == MouseEvent.BUTTON1 && e.getClickCount() == 1;
    }

    private boolean isDoubleLeftClick(MouseEvent e) {
        return e.getButton() == MouseEvent.BUTTON1 && e.getClickCount() == 2;
    }

    private Point calculatePosition(MouseEvent e) {
        Point point = e.getPoint();
        Rectangle bounds = getScreenViewableBounds(getGraphicsDeviceAt(point));
        int x = point.x;
        int y = point.y;

        if (y < bounds.y) {
            y = bounds.y;
        } else if (y > bounds.y + bounds.height) {
            y = bounds.y + bounds.height;
        }
        if (x < bounds.x) {
            x = bounds.x;
        } else if (x > bounds.x + bounds.width) {
            x = bounds.x + bounds.width;
        }
        if (x + popup.getWidth() > bounds.x + bounds.width) {
            x = bounds.x + bounds.width - popup.getWidth();
        }
        if (y + popup.getWidth() > bounds.y + bounds.height) {
            y = bounds.y + bounds.height - popup.getHeight();
        }
        return point;
    }

    private GraphicsDevice getGraphicsDeviceAt(Point pos) {
        GraphicsDevice device = null;
        GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
        GraphicsDevice lstGDs[] = ge.getScreenDevices();
        ArrayList<GraphicsDevice> lstDevices = new ArrayList<>(lstGDs.length);

        for (GraphicsDevice gd : lstGDs) {
            GraphicsConfiguration gc = gd.getDefaultConfiguration();
            Rectangle screenBounds = gc.getBounds();
            if (screenBounds.contains(pos)) {
                lstDevices.add(gd);
            }
        }
        if (lstDevices.size() == 1) {
            device = lstDevices.get(0);
        }
        return device;
    }

    private Rectangle getScreenViewableBounds(GraphicsDevice gd) {
        Rectangle bounds = new Rectangle(0, 0, 0, 0);
        if (gd != null) {
            GraphicsConfiguration gc = gd.getDefaultConfiguration();
            bounds = gc.getBounds();
            Insets insets = Toolkit.getDefaultToolkit().getScreenInsets(gc);

            bounds.x += insets.left;
            bounds.y += insets.top;
            bounds.width -= insets.left + insets.right;
            bounds.height -= insets.top + insets.bottom;
        }
        return bounds;
    }
}

File: client/src/main/java/de/qabel/desktop/ui/tray/TrayIconMouseAdapter.kt
package de.qabel.desktop.ui.tray

import de.qabel.core.extensions.letApply
import java.awt.*
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.event.MouseMotionAdapter
import java.util.*
import javax.swing.JMenuItem
import javax.swing.JPopupMenu

internal class TrayIconMouseAdapter(private val showApp: Runnable, private val bringAppToFront: Runnable) : MouseAdapter() {
    val popup: JPopupMenu
    var popupCloseCheckDelay = 500
    var popupCloseCheckPeriod = 1500
    private var inBound: Boolean = false
    private var notificationTimer: Timer? = null

    init {
        popup = buildSystemTrayJPopupMenu()

        popup.addMouseMotionListener(object : MouseMotionAdapter() {
            override fun mouseMoved(e: MouseEvent?) {
                inBound = true
            }
        })

        popup.addMouseListener(object : MouseAdapter() {
            override fun mouseExited(e: MouseEvent) {
                if (isInBounds(e, popup)) {
                    return
                }
                inBound = false
            }

            private fun isInBounds(e: MouseEvent, popup: JPopupMenu): Boolean {
                return e.x < popup.bounds.maxX &&
                        e.x >= popup.bounds.minX &&
                        e.y < popup.bounds.maxY &&
                        e.y >= popup.bounds.minY
            }
        })
    }

    private fun hideIfOutOfBounds() {
        if (popup.isVisible && !inBound) {
            popup.isVisible = false
            stopTimer()
        }
        inBound = false
    }

    @Synchronized private fun startTimer() {
        stopTimer()
        notificationTimer = Timer().apply {
            schedule(object : TimerTask() {
                override fun run() {
                    hideIfOutOfBounds()
                }
            }, popupCloseCheckDelay.toLong(), popupCloseCheckPeriod.toLong())
        }
    }

    @Synchronized private fun stopTimer() {
        if (notificationTimer != null) {
            notificationTimer!!.cancel()
            notificationTimer = null
        }
    }

    fun setInBound(inBound: Boolean) {
        this.inBound = inBound
    }

    private fun buildSystemTrayJPopupMenu(): JPopupMenu {
        return  JPopupMenu().apply {
            JMenuItem("Exit").letApply {
                it.addActionListener { ae -> System.exit(0) }
                add(it)
            }
        }
    }

    override fun mouseClicked(e: MouseEvent) {
        if (isSingleLeftClick(e)) {
            bringAppToFront.run()
        } else if (isDoubleLeftClick(e)) {
            showApp.run()
        } else if (isRightClick(e)) {
            val position = calculatePosition(e)
            popup.setLocation(position.x, position.y)
            popup.isVisible = true
            startTimer()
        }
    }

    private fun isRightClick(e: MouseEvent): Boolean {
        return e.button == MouseEvent.BUTTON3
    }

    private fun isSingleLeftClick(e: MouseEvent): Boolean {
        return e.button == MouseEvent.BUTTON1 && e.clickCount == 1
    }

    private fun isDoubleLeftClick(e: MouseEvent): Boolean {
        return e.button == MouseEvent.BUTTON1 && e.clickCount == 2
    }

    private fun calculatePosition(e: MouseEvent): Point {
        val point = e.point
        val bounds = getScreenViewableBounds(getGraphicsDeviceAt(point))
        var x = point.x
        var y = point.y

        if (y < bounds.y) {
            y = bounds.y
        } else if (y > bounds.y + bounds.height) {
            y = bounds.y + bounds.height
        }
        if (x < bounds.x) {
            x = bounds.x
        } else if (x > bounds.x + bounds.width) {
            x = bounds.x + bounds.width
        }
        if (x + popup.width > bounds.x + bounds.width) {
            x = bounds.x + bounds.width - popup.width
        }
        if (y + popup.width > bounds.y + bounds.height) {
            y = bounds.y + bounds.height - popup.height
        }
        return point
    }

    private fun getGraphicsDeviceAt(pos: Point): GraphicsDevice {
        val ge = GraphicsEnvironment.getLocalGraphicsEnvironment()
        val lstGDs = ge.screenDevices
        val lstDevices = ArrayList<GraphicsDevice>(lstGDs.size)

        for (gd in lstGDs) {
            val gc = gd.defaultConfiguration
            val screenBounds = gc.bounds
            if (screenBounds.contains(pos)) {
                lstDevices.add(gd)
            }
        }
        return lstDevices.first()
    }

    private fun getScreenViewableBounds(gd: GraphicsDevice?): Rectangle {
        var bounds = Rectangle(0, 0, 0, 0)
        if (gd != null) {
            val gc = gd.defaultConfiguration
            bounds = gc.bounds
            val insets = Toolkit.getDefaultToolkit().getScreenInsets(gc)

            bounds.x += insets.left
            bounds.y += insets.top
            bounds.width -= insets.left + insets.right
            bounds.height -= insets.top + insets.bottom
        }
        return bounds
    }
}
